pyteal/compiler/subroutines_test.py

```diff
@@ -5,21 +5,138 @@
 import pyteal as pt
 
 from pyteal.compiler.subroutines import (
-    findRecursionPoints,
+    find_callstack_exclusive_subroutines,
+    find_recursion_points,
     spillLocalSlotsDuringRecursion,
     resolveSubroutines,
 )
 
 
-def test_findRecursionPoints_empty():
+def test_find_callstack_exclusive_subroutines_empty():
     subroutines = dict()
 
     expected = dict()
-    actual = findRecursionPoints(subroutines)
+    actual = find_callstack_exclusive_subroutines(subroutines)
     assert actual == expected
 
 
-def test_findRecursionPoints_none():
+def test_find_callstack_exclusive_subroutines_none():
+    def sub1Impl():
+        return None
+
+    def sub2Impl(a1):
+        return None
+
+    def sub3Impl(a1, a2, a3):
+        return None
+
+    subroutine1 = pt.SubroutineDefinition(sub1Impl, pt.TealType.uint64)
+    subroutine2 = pt.SubroutineDefinition(sub2Impl, pt.TealType.bytes)
+    subroutine3 = pt.SubroutineDefinition(sub3Impl, pt.TealType.none)
+
+    subroutines = {
+        subroutine1: {subroutine2},
+        subroutine2: {subroutine3},
+        subroutine3: set(),
+    }
+
+    expected = {
+        subroutine1: set(),
+        subroutine2: set(),
+        subroutine3: set(),
+    }
+
+    actual = find_callstack_exclusive_subroutines(subroutines)
+    assert actual == expected
+
+
+def test_find_callstack_exclusive_subroutines_leaf_subroutines():
+    def sub1Impl():
+        return None
+
+    def sub2Impl(a1):
+        return None
+
+    def sub3Impl(a1, a2, a3):
+        return None
+
+    def sub4Impl():
+        return None
+
+    def sub5Impl():
+        return None
+
+    subroutine1 = pt.SubroutineDefinition(sub1Impl, pt.TealType.uint64)
+    subroutine2 = pt.SubroutineDefinition(sub2Impl, pt.TealType.bytes)
+    subroutine3 = pt.SubroutineDefinition(sub3Impl, pt.TealType.none)
+    subroutine4 = pt.SubroutineDefinition(sub4Impl, pt.TealType.none)
+    subroutine5 = pt.SubroutineDefinition(sub5Impl, pt.TealType.none)
+
+    subroutines = {
+        subroutine1: {subroutine2, subroutine4},
+        subroutine2: {subroutine3, subroutine5},
+        subroutine3: set(),
+        subroutine4: set(),
+        subroutine5: set(),
+    }
+
+    expected = {
+        subroutine1: set(),
+        subroutine2: {subroutine4},
+        subroutine3: {subroutine4, subroutine5},
+        subroutine4: {subroutine2, subroutine3, subroutine5},
+        subroutine5: {subroutine3, subroutine4},
+    }
```

Comment:
This approach looks promising though I haven't dug into the details.

Summarizing this test example as a graph (represented by a file-system-like tree hierarchy):

```
s1 -> s2 -> s3
               -> s5
     -> s4
```
and suppose that each subroutine requires exacty 10 slots. Then if my understanding is correct, you end up with the final assignments:
```
s1: 20-29
s2: 10-19
s3: 0-9
s4: 0-9
s5: 0-9
```
This is the same slot assignment as I get from the theoretical pseudocode that I've written elsewhere.
Is my understanding correct?
And if so, what assignment would the following graph produce?

```
s1 -> s2 -> s3 -> s4
               -> s5 -> s6
```
The SCC algo ought to produce:
```
s1: 30-39
s2: 20-29
s3: 10-19
s4: 0-9
s5: 10-19
s6: 0-9
```

Reply:
> And if so, what assignment would the following graph produce?
> 
> ```
> s1 -> s2 -> s3 -> s4
>          -> s5 -> s6
> ```

I expect the code here would produce the same result that you indicated above, or this:

```
s1: 30-39
s2: 20-29
s3: 10-19
s4: 0-9
s5: 0-9
s6: 10-19
```